File: app/src/main/java/com/hiddenramblings/tagmo/browser/adapter/WriteTagAdapter.java
package com.hiddenramblings.tagmo.browser.adapter;

import android.annotation.SuppressLint;
import android.graphics.Bitmap;
import android.graphics.drawable.Drawable;
import android.os.Build;
import android.text.TextUtils;
import android.util.TypedValue;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Filter;
import android.widget.Filterable;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.AppCompatImageView;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.request.target.CustomTarget;
import com.bumptech.glide.request.transition.Transition;
import com.hiddenramblings.tagmo.GlideApp;
import com.hiddenramblings.tagmo.R;
import com.hiddenramblings.tagmo.TagMo;
import com.hiddenramblings.tagmo.amiibo.Amiibo;
import com.hiddenramblings.tagmo.amiibo.AmiiboFile;
import com.hiddenramblings.tagmo.amiibo.AmiiboFileComparator;
import com.hiddenramblings.tagmo.amiibo.AmiiboManager;
import com.hiddenramblings.tagmo.browser.BrowserSettings;
import com.hiddenramblings.tagmo.browser.BrowserSettings.BrowserSettingsListener;
import com.hiddenramblings.tagmo.browser.BrowserSettings.VIEW;
import com.hiddenramblings.tagmo.browser.Preferences;
import com.hiddenramblings.tagmo.eightbit.io.Debug;
import com.hiddenramblings.tagmo.eightbit.os.Storage;
import com.hiddenramblings.tagmo.widget.BoldSpannable;

import java.util.ArrayList;
import java.util.Collections;

public class WriteTagAdapter extends RecyclerView.Adapter<WriteTagAdapter.AmiiboViewHolder>
        implements Filterable, BrowserSettingsListener {

    static Preferences mPrefs = new Preferences(TagMo.getContext());

    private final BrowserSettings settings;
    private OnAmiiboClickListener listener = null;
    private int listSize = 1;
    private ArrayList<AmiiboFile> amiiboFiles = new ArrayList<>();
    private ArrayList<AmiiboFile> filteredData;
    private AmiiboFilter filter;
    boolean firstRun;
    private final ArrayList<AmiiboFile> amiiboList = new ArrayList<>();

    public WriteTagAdapter(BrowserSettings settings) {
        this.settings = settings;

        firstRun = true;
        this.filteredData = this.amiiboFiles;
        this.setHasStableIds(true);
    }

    public void setListener(OnAmiiboClickListener listener, int listSize) {
        this.amiiboList.clear();
        this.listener = listener;
        this.listSize = listSize;
    }

    @Override
    public void onBrowserSettingsChanged(BrowserSettings newBrowserSettings,
                                         BrowserSettings oldBrowserSettings) {
        boolean refresh = firstRun ||
                !BrowserSettings.equals(newBrowserSettings.getQuery(),
                        oldBrowserSettings.getQuery()) ||
                !BrowserSettings.equals(newBrowserSettings.getSort(),
                        oldBrowserSettings.getSort()) ||
                BrowserSettings.hasFilterChanged(oldBrowserSettings, newBrowserSettings);
        
        if (firstRun || !BrowserSettings.equals(newBrowserSettings.getAmiiboFiles(),
                oldBrowserSettings.getAmiiboFiles())) {
            this.amiiboFiles = new ArrayList<>();
            if (null != newBrowserSettings.getAmiiboFiles())
                this.amiiboFiles.addAll(newBrowserSettings.getAmiiboFiles());
            refresh = true;
        }

        if (!BrowserSettings.equals(newBrowserSettings.getAmiiboManager(),
                oldBrowserSettings.getAmiiboManager())) {
            refresh = true;
        }
        if (!BrowserSettings.equals(newBrowserSettings.getAmiiboView(),
                oldBrowserSettings.getAmiiboView())) {
            refresh = true;
        }
        if (refresh) {
            this.refresh();
        }
    }

    @Override
    public int getItemCount() {
        return filteredData.size();
    }

    @Override
    public long getItemId(int i) {
        return filteredData.get(i).getId();
    }

    public AmiiboFile getItem(int i) {
        return filteredData.get(i);
    }

    @Override
    public int getItemViewType(int position) {
        return settings.getAmiiboView();
    }

    @NonNull
    @Override
    public AmiiboViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        switch (VIEW.valueOf(viewType)) {
            case COMPACT:
                return new CompactViewHolder(parent, settings);
            case LARGE:
                return new LargeViewHolder(parent, settings);
            case IMAGE:
                return new ImageViewHolder(parent, settings);
            case SIMPLE:
            default:
                return new SimpleViewHolder(parent, settings);
        }
    }

    private void setIsHighlighted(AmiiboViewHolder holder, boolean isHighlighted) {
        View highlight = holder.itemView.findViewById(R.id.highlight);
        if (isHighlighted) {
            highlight.setBackgroundResource(R.drawable.rounded_neon);
        } else {
            highlight.setBackgroundResource(0);
        }
    }

    private void handleClickEvent(final AmiiboViewHolder holder, int position) {
        if (null != listener) {
            if (listSize > 1) {
                if (amiiboList.contains(holder.amiiboFile)) {
                    amiiboList.remove(filteredData.get(position));
                    setIsHighlighted(holder, false);
                } else {
                    amiiboList.add(filteredData.get(position));
                    setIsHighlighted(holder, true);
                }
                if (amiiboList.size() == listSize)
                    listener.onAmiiboListClicked(amiiboList);
            } else {
                listener.onAmiiboClicked(holder.amiiboFile);
            }
        }
    }

    @Override
    public void onBindViewHolder(@NonNull final AmiiboViewHolder holder, int position) {
        final int clickPosition = hasStableIds() ? holder.getBindingAdapterPosition() : position;
        holder.itemView.setOnClickListener(view -> handleClickEvent(holder, clickPosition));
        if (null != holder.imageAmiibo) {
            holder.imageAmiibo.setOnClickListener(view -> {
                if (settings.getAmiiboView() == VIEW.IMAGE.getValue())
                    handleClickEvent(holder, clickPosition);
                else if (null != listener)
                    listener.onAmiiboImageClicked(holder.amiiboFile);
            });
        }
        holder.bind(getItem(clickPosition));
        setIsHighlighted(holder, amiiboList.contains(holder.amiiboFile));
    }

    public void refresh() {
        this.getFilter().filter(settings.getQuery());
    }

    @Override
    public AmiiboFilter getFilter() {
        if (null == this.filter) {
            this.filter = new AmiiboFilter();
        }
        return this.filter;
    }

    class AmiiboFilter extends Filter {
        @Override
        protected FilterResults performFiltering(CharSequence constraint) {
            FilterResults filterResults = new FilterResults();
            ArrayList<AmiiboFile> tempList = new ArrayList<>();
            String queryText = settings.getQuery().trim().toLowerCase();
            AmiiboManager amiiboManager = settings.getAmiiboManager();
            for (AmiiboFile amiiboFile : amiiboFiles) {
                boolean add = false;
                if (null != amiiboManager) {
                    Amiibo amiibo = amiiboManager.amiibos.get(amiiboFile.getId());
                    if (null == amiibo)
                        amiibo = new Amiibo(amiiboManager, amiiboFile.getId(), null, null);
                    add = settings.amiiboContainsQuery(amiibo, queryText);
                }
                if (!add && null != amiiboFile.getDocUri())
                    add = pathContainsQuery(amiiboFile.getDocUri().toString(), queryText);
                if (!add && null != amiiboFile.getFilePath())
                    add = pathContainsQuery(amiiboFile.getFilePath().getAbsolutePath(), queryText);
                if (add)
                    tempList.add(amiiboFile);
            }
            filterResults.count = tempList.size();
            filterResults.values = tempList;

            return filterResults;
        }

        public boolean pathContainsQuery(String path, String query) {
            return !TextUtils.isEmpty(query) && settings.isFilterEmpty()
                    && path.toLowerCase().contains(query);
        }

        @SuppressLint("NotifyDataSetChanged")
        @Override
        protected void publishResults(CharSequence charSequence, FilterResults filterResults) {
            if (null != filteredData && filteredData == filterResults.values) return;
            //noinspection unchecked
            filteredData = (ArrayList<AmiiboFile>) filterResults.values;
            if (null != filteredData && null != settings)
                Collections.sort(filteredData, new AmiiboFileComparator(settings));
            notifyDataSetChanged();
        }
    }

    protected static abstract class AmiiboViewHolder extends RecyclerView.ViewHolder {
        private final BrowserSettings settings;

        public final TextView txtError;
        public final TextView txtName;
        public final TextView txtTagId;
        public final TextView txtAmiiboSeries;
        public final TextView txtAmiiboType;
        public final TextView txtGameSeries;
        // public final TextView txtCharacter;
        public final TextView txtPath;
        public final AppCompatImageView imageAmiibo;

        private AmiiboFile amiiboFile = null;

        private final BoldSpannable boldSpannable = new BoldSpannable();

        private final CustomTarget<Bitmap> target = new CustomTarget<>() {
            @Override
            public void onLoadStarted(@Nullable Drawable placeholder) {
                imageAmiibo.setImageResource(0);
            }

            @Override
            public void onLoadFailed(@Nullable Drawable errorDrawable) {
                imageAmiibo.setVisibility(View.INVISIBLE);
            }

            @Override
            public void onLoadCleared(@Nullable Drawable placeholder) {
                imageAmiibo.setVisibility(View.VISIBLE);
            }

            @Override
            public void onResourceReady(@NonNull Bitmap resource, Transition transition) {
                imageAmiibo.setImageBitmap(resource);
                imageAmiibo.setVisibility(View.VISIBLE);
            }
        };

        public AmiiboViewHolder(View itemView, BrowserSettings settings) {
            super(itemView);

            this.settings = settings;

            this.txtError = itemView.findViewById(R.id.txtError);
            this.txtName = itemView.findViewById(R.id.txtName);
            this.txtTagId = itemView.findViewById(R.id.txtTagId);
            this.txtAmiiboSeries = itemView.findViewById(R.id.txtAmiiboSeries);
            this.txtAmiiboType = itemView.findViewById(R.id.txtAmiiboType);
            this.txtGameSeries = itemView.findViewById(R.id.txtGameSeries);
            // this.txtCharacter = itemView.findViewById(R.id.txtCharacter);
            this.txtPath = itemView.findViewById(R.id.txtPath);
            this.imageAmiibo = itemView.findViewById(R.id.imageAmiibo);
        }

        void bind(final AmiiboFile item) {
            this.amiiboFile = item;

            String tagInfo = null;
            String amiiboHexId = "";
            String amiiboName = "";
            String amiiboSeries = "";
            String amiiboType = "";
            String gameSeries = "";
            // String character = "";
            String amiiboImageUrl;

            long amiiboId = item.getId();
            Amiibo amiibo = null;
            AmiiboManager amiiboManager = settings.getAmiiboManager();
            if (null != amiiboManager) {
                amiibo = amiiboManager.amiibos.get(amiiboId);
                if (null == amiibo)
                    amiibo = new Amiibo(amiiboManager, amiiboId, null, null);
            }
            if (null != amiibo) {
                amiiboHexId = Amiibo.idToHex(amiibo.id);
                amiiboImageUrl = amiibo.getImageUrl();
                if (null != amiibo.name )
                    amiiboName = amiibo.name;
                if (null != amiibo.getAmiiboSeries() )
                    amiiboSeries = amiibo.getAmiiboSeries().name;
                if (null != amiibo.getAmiiboType() )
                    amiiboType = amiibo.getAmiiboType().name;
                if (null != amiibo.getGameSeries() )
                    gameSeries = amiibo.getGameSeries().name;
            } else {
                tagInfo = "ID: " + Amiibo.idToHex(amiiboId);
                amiiboImageUrl = Amiibo.getImageUrl(amiiboId);
            }

            String query = settings.getQuery().toLowerCase();

            setAmiiboInfoText(this.txtName, amiiboName, false);
            if (settings.getAmiiboView() != VIEW.IMAGE.getValue()) {
                boolean hasTagInfo = null != tagInfo;
                if (hasTagInfo) {
                    setAmiiboInfoText(this.txtError, tagInfo, false);
                } else {
                    this.txtError.setVisibility(View.GONE);
                }
                setAmiiboInfoText(this.txtTagId, boldSpannable.startsWith(amiiboHexId, query), hasTagInfo);
                setAmiiboInfoText(this.txtAmiiboSeries,
                        boldSpannable.indexOf(amiiboSeries, query), hasTagInfo);
                setAmiiboInfoText(this.txtAmiiboType,
                        boldSpannable.indexOf(amiiboType, query), hasTagInfo);
                setAmiiboInfoText(this.txtGameSeries,
                        boldSpannable.indexOf(gameSeries, query), hasTagInfo);
                if (null != item.getDocUri()) {
                    this.itemView.setEnabled(true);
                    String relativeDocument = Storage.getRelativeDocument(
                            item.getDocUri().getUri()
                    );
                    this.txtPath.setText(boldSpannable.indexOf(relativeDocument, query));
                    TypedValue a = new TypedValue();
                    this.txtPath.getContext().getTheme().resolveAttribute(
                            android.R.attr.textColor, a, true);
                    if (Debug.INSTANCE.isNewer(Build.VERSION_CODES.Q) && a.isColorType()) {
                        this.txtPath.setTextColor(a.data);
                    } else if (a.type >= TypedValue.TYPE_FIRST_COLOR_INT
                            && a.type <= TypedValue.TYPE_LAST_COLOR_INT) {
                        this.txtPath.setTextColor(a.data);
                    }
                } else if (null != item.getFilePath()) {
                    this.itemView.setEnabled(true);
                    String relativeFile = Storage.getRelativePath(item.getFilePath(),
                            mPrefs.preferEmulated()).replace(
                            mPrefs.browserRootFolder(), "");
                    this.txtPath.setText(boldSpannable.indexOf(relativeFile, query));
                    TypedValue a = new TypedValue();
                    this.txtPath.getContext().getTheme().resolveAttribute(
                            android.R.attr.textColor, a, true);
                    if (Debug.INSTANCE.isNewer(Build.VERSION_CODES.Q) && a.isColorType()) {
                        this.txtPath.setTextColor(a.data);
                    } else if (a.type >= TypedValue.TYPE_FIRST_COLOR_INT
                            && a.type <= TypedValue.TYPE_LAST_COLOR_INT) {
                        this.txtPath.setTextColor(a.data);
                    }
                } else {
                    this.itemView.setEnabled(false);
                    this.txtPath.setText("");
                    this.txtPath.setTextColor(this.txtPath.getResources().getColor(R.color.tag_text));
                }
                this.txtPath.setVisibility(View.VISIBLE);
            }
            if (null != this.imageAmiibo) {
                GlideApp.with(imageAmiibo).clear(imageAmiibo);
                if (null != amiiboImageUrl) {
                    GlideApp.with(imageAmiibo).asBitmap().load(amiiboImageUrl).into(target);
                }
            }
        }

        void setAmiiboInfoText(TextView textView, CharSequence text, boolean hasTagInfo) {
            if (hasTagInfo) {
                textView.setVisibility(View.GONE);
            } else {
                textView.setVisibility(View.VISIBLE);
                if (text.length() == 0) {
                    textView.setText(R.string.unknown);
                    textView.setEnabled(false);
                } else {
                    textView.setText(text);
                    textView.setEnabled(true);
                }
            }
        }
    }

    static class SimpleViewHolder extends AmiiboViewHolder {
        public SimpleViewHolder(ViewGroup parent, BrowserSettings settings) {
            super(
                    LayoutInflater.from(parent.getContext()).inflate(
                            R.layout.amiibo_simple_card, parent, false),
                    settings
            );
        }
    }

    static class CompactViewHolder extends AmiiboViewHolder {
        public CompactViewHolder(ViewGroup parent, BrowserSettings settings) {
            super(
                    LayoutInflater.from(parent.getContext()).inflate(
                            R.layout.amiibo_compact_card, parent, false),
                    settings
            );
        }
    }

    static class LargeViewHolder extends AmiiboViewHolder {
        public LargeViewHolder(ViewGroup parent, BrowserSettings settings) {
            super(
                    LayoutInflater.from(parent.getContext()).inflate(
                            R.layout.amiibo_large_card, parent, false),
                    settings
            );
        }
    }

    static class ImageViewHolder extends AmiiboViewHolder {
        public ImageViewHolder(ViewGroup parent, BrowserSettings settings) {
            super(
                    LayoutInflater.from(parent.getContext()).inflate(
                            R.layout.amiibo_image_card, parent, false),
                    settings
            );
        }
    }

    public interface OnAmiiboClickListener {
        void onAmiiboClicked(AmiiboFile amiiboFile);
        void onAmiiboImageClicked(AmiiboFile amiiboFile);
        void onAmiiboListClicked(ArrayList<AmiiboFile> amiiboList);
    }
}

File: app/src/main/java/com/hiddenramblings/tagmo/browser/adapter/WriteTagAdapter.kt
package com.hiddenramblings.tagmo.browser.adapter

import android.annotation.SuppressLint
import android.graphics.Bitmap
import android.graphics.drawable.Drawable
import android.os.Build
import android.text.TextUtils
import android.util.TypedValue
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Filter
import android.widget.Filterable
import android.widget.TextView
import androidx.appcompat.widget.AppCompatImageView
import androidx.core.content.ContextCompat
import androidx.core.view.isGone
import androidx.core.view.isVisible
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.request.target.CustomTarget
import com.bumptech.glide.request.transition.Transition
import com.hiddenramblings.tagmo.GlideApp
import com.hiddenramblings.tagmo.Preferences
import com.hiddenramblings.tagmo.R
import com.hiddenramblings.tagmo.TagMo
import com.hiddenramblings.tagmo.amiibo.Amiibo
import com.hiddenramblings.tagmo.amiibo.AmiiboFile
import com.hiddenramblings.tagmo.amiibo.AmiiboFileComparator
import com.hiddenramblings.tagmo.browser.BrowserSettings
import com.hiddenramblings.tagmo.browser.BrowserSettings.BrowserSettingsListener
import com.hiddenramblings.tagmo.browser.BrowserSettings.VIEW
import com.hiddenramblings.tagmo.eightbit.io.Debug.isNewer
import com.hiddenramblings.tagmo.eightbit.os.Storage
import com.hiddenramblings.tagmo.widget.BoldSpannable
import java.util.*

class WriteTagAdapter(private val settings: BrowserSettings?) :
    RecyclerView.Adapter<WriteTagAdapter.AmiiboViewHolder>(), Filterable, BrowserSettingsListener {
    private var listener: OnAmiiboClickListener? = null
    private var listSize = 1
    private var amiiboFiles = ArrayList<AmiiboFile?>()
    private var filteredData: ArrayList<AmiiboFile?>?
    private var filter: AmiiboFilter? = null
    private var firstRun = true
    private val amiiboList = ArrayList<AmiiboFile?>()

    init {
        filteredData = amiiboFiles
        setHasStableIds(true)
    }

    fun setListener(listener: OnAmiiboClickListener?, listSize: Int) {
        amiiboList.clear()
        this.listener = listener
        this.listSize = listSize
    }

    override fun onBrowserSettingsChanged(
        newBrowserSettings: BrowserSettings?,
        oldBrowserSettings: BrowserSettings?
    ) {
        var refresh = firstRun ||
                !BrowserSettings.equals(
                    newBrowserSettings?.query,
                    oldBrowserSettings?.query
                ) ||
                !BrowserSettings.equals(
                    newBrowserSettings?.sort,
                    oldBrowserSettings?.sort
                ) ||
                BrowserSettings.hasFilterChanged(oldBrowserSettings, newBrowserSettings)
        if (firstRun || !BrowserSettings.equals(
                newBrowserSettings?.amiiboFiles,
                oldBrowserSettings?.amiiboFiles
            )
        ) {
            amiiboFiles = ArrayList()
            if (null != newBrowserSettings?.amiiboFiles)
                amiiboFiles.addAll(newBrowserSettings.amiiboFiles)
            refresh = true
        }
        if (!BrowserSettings.equals(
                newBrowserSettings?.amiiboManager,
                oldBrowserSettings?.amiiboManager
            )
        ) {
            refresh = true
        }
        if (!BrowserSettings.equals(
                newBrowserSettings?.amiiboView,
                oldBrowserSettings?.amiiboView
            )
        ) {
            refresh = true
        }
        if (refresh) {
            refresh()
        }
    }

    override fun getItemCount(): Int {
        return filteredData?.size ?: 0
    }

    override fun getItemId(i: Int): Long {
        return (filteredData?.get(i)?.id ?: 0).toLong()
    }

    private fun getItem(i: Int): AmiiboFile? {
        return filteredData?.get(i)
    }

    override fun getItemViewType(position: Int): Int {
        return settings?.amiiboView ?: 0
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): AmiiboViewHolder {
        return when (VIEW.valueOf(viewType)) {
            VIEW.COMPACT -> CompactViewHolder(parent, settings)
            VIEW.LARGE -> LargeViewHolder(parent, settings)
            VIEW.IMAGE -> ImageViewHolder(parent, settings)
            VIEW.SIMPLE -> SimpleViewHolder(parent, settings)
        }
    }

    private fun setIsHighlighted(holder: AmiiboViewHolder, isHighlighted: Boolean) {
        val highlight = holder.itemView.findViewById<View>(R.id.highlight)
        if (isHighlighted) {
            highlight.setBackgroundResource(R.drawable.rounded_neon)
        } else {
            highlight.setBackgroundResource(0)
        }
    }

    private fun handleClickEvent(holder: AmiiboViewHolder, position: Int) {
        if (listSize > 1) {
            if (amiiboList.contains(holder.amiiboFile)) {
                amiiboList.remove(filteredData?.get(position))
                setIsHighlighted(holder, false)
            } else {
                amiiboList.add(filteredData?.get(position))
                setIsHighlighted(holder, true)
            }
            if (amiiboList.size == listSize) listener?.onAmiiboListClicked(amiiboList)
        } else {
            listener?.onAmiiboClicked(holder.amiiboFile)
        }
    }

    override fun onBindViewHolder(holder: AmiiboViewHolder, position: Int) {
        val clickPosition = if (hasStableIds()) holder.bindingAdapterPosition else position
        holder.itemView.setOnClickListener {
            handleClickEvent(holder, clickPosition)
        }
        holder.imageAmiibo?.setOnClickListener {
            if (settings!!.amiiboView == VIEW.IMAGE.value)
                handleClickEvent(holder, clickPosition)
            else if (null != listener)
                listener!!.onAmiiboImageClicked(holder.amiiboFile)
        }
        holder.bind(getItem(clickPosition)!!)
        setIsHighlighted(holder, amiiboList.contains(holder.amiiboFile))
    }

    fun refresh() {
        getFilter().filter(settings?.query)
    }

    override fun getFilter(): AmiiboFilter {
        if (null == filter) {
            filter = AmiiboFilter()
        }
        return filter!!
    }

    inner class AmiiboFilter : Filter() {
        override fun performFiltering(constraint: CharSequence): FilterResults {
            val filterResults = FilterResults()
            val tempList = ArrayList<AmiiboFile>()
            val queryText = settings?.query!!.trim { it <= ' ' }.lowercase(Locale.getDefault())
            val amiiboManager = settings.amiiboManager
            amiiboFiles.forEach {
                var add = false
                if (null != amiiboManager) {
                    var amiibo = amiiboManager.amiibos[it?.id]
                    if (null == amiibo) amiibo = Amiibo(amiiboManager, it!!.id, null, null)
                    add = settings.amiiboContainsQuery(amiibo, queryText)
                }
                if (!add && null != it?.docUri) add =
                    pathContainsQuery(it.docUri.toString(), queryText)
                if (!add && null != it?.filePath) add =
                    pathContainsQuery(it.filePath!!.absolutePath, queryText)
                if (add) tempList.add(it!!)
            }
            filterResults.count = tempList.size
            filterResults.values = tempList
            return filterResults
        }

        private fun pathContainsQuery(path: String, query: String?): Boolean {
            return (!TextUtils.isEmpty(query) && settings!!.isFilterEmpty
                    && path.lowercase(Locale.getDefault()).contains(query!!))
        }

        @SuppressLint("NotifyDataSetChanged")
        override fun publishResults(charSequence: CharSequence?, filterResults: FilterResults) {
            if (null != filteredData && filteredData === filterResults.values) return
            filteredData = filterResults.values as ArrayList<AmiiboFile?>
            if (null != filteredData && null != settings)
                Collections.sort(filteredData, AmiiboFileComparator(settings))
            notifyDataSetChanged()
        }
    }

    abstract class AmiiboViewHolder(
        itemView: View, private val settings: BrowserSettings?
    ) : RecyclerView.ViewHolder(itemView) {
        val txtError: TextView
        val txtName: TextView
        val txtTagId: TextView
        val txtAmiiboSeries: TextView
        val txtAmiiboType: TextView
        val txtGameSeries: TextView

        // public final TextView txtCharacter;
        val txtPath: TextView
        var imageAmiibo: AppCompatImageView? = null
        var amiiboFile: AmiiboFile? = null
        private val boldSpannable = BoldSpannable()
        private val target: CustomTarget<Bitmap?> = object : CustomTarget<Bitmap?>() {
            override fun onLoadStarted(placeholder: Drawable?) {
                imageAmiibo?.setImageResource(0)
            }

            override fun onLoadFailed(errorDrawable: Drawable?) {
                imageAmiibo?.visibility = View.INVISIBLE
            }

            override fun onLoadCleared(placeholder: Drawable?) {
                imageAmiibo?.visibility = View.VISIBLE
            }

            override fun onResourceReady(resource: Bitmap, transition: Transition<in Bitmap?>?) {
                imageAmiibo?.setImageBitmap(resource)
                imageAmiibo?.visibility = View.VISIBLE
            }
        }

        init {
            txtError = itemView.findViewById(R.id.txtError)
            txtName = itemView.findViewById(R.id.txtName)
            txtTagId = itemView.findViewById(R.id.txtTagId)
            txtAmiiboSeries = itemView.findViewById(R.id.txtAmiiboSeries)
            txtAmiiboType = itemView.findViewById(R.id.txtAmiiboType)
            txtGameSeries = itemView.findViewById(R.id.txtGameSeries)
            // this.txtCharacter = itemView.findViewById(R.id.txtCharacter);
            txtPath = itemView.findViewById(R.id.txtPath)
            imageAmiibo = itemView.findViewById(R.id.imageAmiibo)
        }

        fun bind(item: AmiiboFile?) {
            amiiboFile = item
            var tagInfo: String? = null
            var amiiboHexId: String? = ""
            var amiiboName = ""
            var amiiboSeries = ""
            var amiiboType = ""
            var gameSeries = ""
            // String character = "";
            var amiiboImageUrl: String? = null
            val amiiboId = item?.id
            var amiibo: Amiibo? = null
            val amiiboManager = settings!!.amiiboManager
            if (null != amiiboManager) {
                amiibo = amiiboManager.amiibos[amiiboId]
                if (null == amiibo && null != amiiboId)
                    amiibo = Amiibo(amiiboManager, amiiboId, null, null)
            }
            if (null != amiibo) {
                amiiboHexId = Amiibo.idToHex(amiibo.id)
                amiiboImageUrl = amiibo.imageUrl
                if (null != amiibo.name) amiiboName = amiibo.name!!
                if (null != amiibo.amiiboSeries) amiiboSeries = amiibo.amiiboSeries!!.name
                if (null != amiibo.amiiboType) amiiboType = amiibo.amiiboType!!.name
                if (null != amiibo.gameSeries) gameSeries = amiibo.gameSeries!!.name
            } else if (null != amiiboId) {
                tagInfo = "ID: " + Amiibo.idToHex(amiiboId)
                amiiboImageUrl = Amiibo.getImageUrl(amiiboId)
            }
            val query = settings.query?.lowercase(Locale.getDefault())
            setAmiiboInfoText(txtName, amiiboName, false)
            if (settings.amiiboView != VIEW.IMAGE.value) {
                val hasTagInfo = null != tagInfo
                if (hasTagInfo) {
                    setAmiiboInfoText(txtError, tagInfo, false)
                } else {
                    txtError.isGone = true
                }
                setAmiiboInfoText(
                    txtTagId,
                    boldSpannable.startsWith(amiiboHexId, query),
                    hasTagInfo
                )
                setAmiiboInfoText(
                    txtAmiiboSeries,
                    boldSpannable.indexOf(amiiboSeries, query), hasTagInfo
                )
                setAmiiboInfoText(
                    txtAmiiboType,
                    boldSpannable.indexOf(amiiboType, query), hasTagInfo
                )
                setAmiiboInfoText(
                    txtGameSeries,
                    boldSpannable.indexOf(gameSeries, query), hasTagInfo
                )
                if (null != item?.docUri) {
                    itemView.isEnabled = true
                    val relativeDocument = Storage.getRelativeDocument(
                        item.docUri!!.uri
                    )
                    txtPath.text = boldSpannable.indexOf(relativeDocument, query)
                    val a = TypedValue()
                    txtPath.context.theme.resolveAttribute(
                        android.R.attr.textColor, a, true
                    )
                    if (isNewer(Build.VERSION_CODES.Q) && a.isColorType) {
                        txtPath.setTextColor(a.data)
                    } else if (a.type >= TypedValue.TYPE_FIRST_COLOR_INT
                        && a.type <= TypedValue.TYPE_LAST_COLOR_INT
                    ) {
                        txtPath.setTextColor(a.data)
                    }
                } else if (null != item?.filePath) {
                    itemView.isEnabled = true
                    var relativeFile = Storage.getRelativePath(
                        item.filePath, mPrefs.preferEmulated()
                    )
                    if (null != mPrefs.browserRootFolder())
                        relativeFile = relativeFile.replace(mPrefs.browserRootFolder()!!, "")
                    txtPath.text = boldSpannable.indexOf(relativeFile, query)
                    val a = TypedValue()
                    txtPath.context.theme.resolveAttribute(
                        android.R.attr.textColor, a, true
                    )
                    if (isNewer(Build.VERSION_CODES.Q) && a.isColorType) {
                        txtPath.setTextColor(a.data)
                    } else if (a.type >= TypedValue.TYPE_FIRST_COLOR_INT
                        && a.type <= TypedValue.TYPE_LAST_COLOR_INT
                    ) {
                        txtPath.setTextColor(a.data)
                    }
                } else {
                    itemView.isEnabled = false
                    txtPath.text = ""
                    txtPath.setTextColor(ContextCompat.getColor(txtPath.context, R.color.tag_text))
                }
                txtPath.isVisible = true
            }
            if (null != imageAmiibo) {
                GlideApp.with(imageAmiibo!!).clear(imageAmiibo!!)
                if (!amiiboImageUrl.isNullOrEmpty()) {
                    GlideApp.with(imageAmiibo!!).asBitmap().load(amiiboImageUrl).into(target)
                }
            }
        }

        fun setAmiiboInfoText(textView: TextView, text: CharSequence?, hasTagInfo: Boolean) {
            textView.isGone = hasTagInfo
            if (!hasTagInfo) {
                 if (!text.isNullOrEmpty()) {
                    textView.text = text
                    textView.isEnabled = true
                } else {
                    textView.setText(R.string.unknown)
                    textView.isEnabled = false
                }
            }
        }
    }

    internal class SimpleViewHolder(parent: ViewGroup, settings: BrowserSettings?) :
        AmiiboViewHolder(
            LayoutInflater.from(parent.context).inflate(
                R.layout.amiibo_simple_card, parent, false
            ),
            settings
        )

    internal class CompactViewHolder(parent: ViewGroup, settings: BrowserSettings?) :
        AmiiboViewHolder(
            LayoutInflater.from(parent.context).inflate(
                R.layout.amiibo_compact_card, parent, false
            ),
            settings
        )

    internal class LargeViewHolder(parent: ViewGroup, settings: BrowserSettings?) :
        AmiiboViewHolder(
            LayoutInflater.from(parent.context).inflate(
                R.layout.amiibo_large_card, parent, false
            ),
            settings
        )

    internal class ImageViewHolder(parent: ViewGroup, settings: BrowserSettings?) :
        AmiiboViewHolder(
            LayoutInflater.from(parent.context).inflate(
                R.layout.amiibo_image_card, parent, false
            ),
            settings
        )

    interface OnAmiiboClickListener {
        fun onAmiiboClicked(amiiboFile: AmiiboFile?)
        fun onAmiiboImageClicked(amiiboFile: AmiiboFile?)
        fun onAmiiboListClicked(amiiboList: ArrayList<AmiiboFile?>?)
    }

    companion object {
        var mPrefs = Preferences(TagMo.appContext)
    }
}